File: baekjoon/java/src/silver2/Q17829_222Pooling.java
package silver2;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;

// https://www.acmicpc.net/problem/17829
public class Q17829_222Pooling {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] board = new int[n][n];
        for (int i = 0; i < n; i++) {
            String[] data = br.readLine().split(" ");
            for (int j = 0; j < data.length; j++) {
                board[i][j] = Integer.parseInt(data[j]);
            }
        }

        while (n > 1) {
            n /= 2;
            int[][] newBoard = new int[n][n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    Integer[] toCompare = {board[i * 2][j * 2], board[i * 2][j * 2 + 1], board[i * 2 + 1][j * 2], board[i * 2 + 1][j * 2 + 1]};
                    Arrays.sort(toCompare, Comparator.reverseOrder());
                    newBoard[i][j] = toCompare[2];
                }
            }
            board = newBoard;
        }

        System.out.println(board[0][0]);
    }
}

File: baekjoon/python/silver2/q17829_222_pooling.py
import sys
input = sys.stdin.readline

# https://www.acmicpc.net/problem/17829
def solution():
    n = int(input())
    board = []
    for _ in range(n):
        board.append(list(map(int, input().split())))

    while n > 2 : 
        n, board = change(n, board)
    
    answer = []
    for i in range(2):
        for j in range(2):
            answer.append(board[i][j])
    answer.sort(reverse=True)
    return answer[1]

def change(n, board):
    idxs = [ i for i in range(0, n-1, 2)]
    new_board = [] 
    for i, x in enumerate(idxs):
        temp1 = []
        for y in idxs:
            temp2 = []
            temp2.append(board[x][y])
            temp2.append(board[x][y+1])
            temp2.append(board[x+1][y+1])
            temp2.append(board[x+1][y])
            temp2.sort(reverse=True)
            temp1.append(temp2[1])
        new_board.append(temp1)
    return n//2, new_board
        
print(solution())


n=int(input())
m=[[*map(int,input().split())]for i in range(n)]
while n>1:
    n//=2
    m2=[n*[0]for i in range(n)]
    for i in range(n):
        for j in range(n):
            m2[i][j]=sorted([m[2*i][2*j],m[2*i][2*j+1],m[2*i+1][2*j],m[2*i+1][2*j+1]])[2]
    m=m2
    print(m2)
print(*m[0])
